### PR TITLE
Add state archival writer and uploader

### PR DESCRIPTION
**Pull request by sadhansood**
## Description 

PR adds archival writer and uploader code. Will add some tests to this soon.

### CODE REVIEW

File: crates/sui-archival/src/writer.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+                    .get_full_checkpoint_contents(&checkpoint_summary.content_digest)
+                    .map_err(|_| anyhow!("Failed to read checkpoint content"))?
+                {
+                    checkpoint_writer
+                        .write(checkpoint_contents, checkpoint_summary.into_inner())
+                        .await?;
+                    return Ok(());
+                }
+            }
+            tokio::time::sleep(Duration::from_secs(3)).await;
```
Comment (mystenmark) [inline crates/sui-archival/src/writer.rs:405]: why sleep?

File: crates/sui-storage/src/lib.rs [hunk 2]
```diff
@@ -2,10 +2,134 @@
...
+        let blob = Blob {
+            data,
+            encoding: Encoding::try_from(encoding)?,
+        };
+        Ok(blob)
+    }
+    pub fn write<W: Write>(&self, wbuf: &mut BufWriter<W>) -> Result<usize> {
+        let mut buf = [0u8; MAX_VARINT_LENGTH];
+        let mut counter = 0;
+        let n = (self.data.len() as u64).encode_var(&mut buf);
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-storage/src/lib.rs:127]: i suppose u32::MAX is probably enough, but why not just support u64?

File: crates/sui-archival/src/writer.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+        let mut f = File::create(next_file_tmp_path.clone())?;
+        let mut metab = [0u8; MAGIC_BYTES];
+        BigEndian::write_u32(&mut metab, magic_bytes);
+        f.rewind()?;
+        let n = f.write(&metab)?;
+        drop(f);
+        fs::rename(next_file_tmp_path, next_file_path.clone())?;
+        let mut f = OpenOptions::new().append(true).open(next_file_path)?;
+        f.seek(SeekFrom::Start(n as u64))?;
+        Ok(f)
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-archival/src/writer.rs:229]: > Just opening a new file with truncation and append doesn't work, I tried it

i'm curious as to why, but i'm also happy to take your word for it

**Review by mystenmark — APPROVED** [review]

File: crates/sui-storage/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadhansood) [reply crates/sui-storage/src/lib.rs:127]: I just thought we would never really need such a big single blob but I'll make it u64::Max for now

**Pull request merged**